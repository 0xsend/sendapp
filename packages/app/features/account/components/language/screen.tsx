import { Card, Paragraph, Separator, Spinner, XStack, YStack } from '@my/ui'
import { SettingsHeader } from 'app/features/account/components/SettingsHeader'
import { useTranslation } from 'react-i18next'
import { useCallback, useMemo, useState } from 'react'
import { Pressable, Platform } from 'react-native'
import { Check } from '@tamagui/lucide-icons'
import { getSupportedLocaleOptions } from 'app/i18n/locales'
import { useHoverStyles } from 'app/utils/useHoverStyles'

export const LanguagePreferences = () => {
  const { i18n, t } = useTranslation('settings')
  const [pendingLocale, setPendingLocale] = useState<string | null>(null)
  const resolvedLanguage = i18n.resolvedLanguage ?? i18n.language

  const options = useMemo(() => getSupportedLocaleOptions(), [])
  const hoverStyles = useHoverStyles()

  const handleSelect = useCallback(
    async (locale: string) => {
      if (locale === resolvedLanguage || pendingLocale === locale) {
        return
      }

      setPendingLocale(locale)
      try {
        await i18n.changeLanguage(locale)
      } finally {
        setPendingLocale(null)
      }
    },
    [i18n, pendingLocale, resolvedLanguage]
  )

  return (
    <YStack gap={'$3.5'} w={'100%'}>
      {Platform.OS === 'web' && <SettingsHeader>{t('language.title')}</SettingsHeader>}
      <Card gap={'$3'} padded size={'$4'}>
        <Paragraph size={'$4'} color={'$color12'}>
          {t('language.description')}
        </Paragraph>
        {pendingLocale ? (
          <Paragraph size={'$2'} color={'$lightGrayTextField'}>
            {t('language.updating')}
          </Paragraph>
        ) : null}
        <Separator boc={'$darkGrayTextField'} opacity={0.2} />
        <YStack>
          {options.map((option, index) => {
            const isActive = option.value === resolvedLanguage
            const isPending = pendingLocale === option.value

            return (
              <YStack key={option.value}>
                <Pressable
                  accessibilityRole="button"
                  accessibilityState={{ selected: isActive, busy: isPending }}
                  disabled={isPending}
                  onPress={() => handleSelect(option.value)}
                >
                  <XStack
                    ai="center"
                    jc="space-between"
                    px={'$2'}
                    py={'$3'}
                    br={'$4'}
                    hoverStyle={hoverStyles}
                  >
                    <Paragraph size={'$5'}>{option.autonym}</Paragraph>
                    {isPending ? (
                      <Spinner size="small" />
                    ) : isActive ? (
                      <Check size={16} color={'$primary'} />
                    ) : null}
                  </XStack>
                </Pressable>
                {index < options.length - 1 && (
                  <Separator boc={'$darkGrayTextField'} opacity={0.15} />
                )}
              </YStack>
            )
          })}
        </YStack>
      </Card>
    </YStack>
  )
}
