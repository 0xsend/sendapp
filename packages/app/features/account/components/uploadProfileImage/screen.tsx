import { SizableText, YStack } from '@my/ui'
import { IconRefresh } from 'app/components/icons'
import { useSupabase } from 'app/utils/supabase/useSupabase'
import { useUser } from 'app/utils/useUser'
import { decode } from 'base64-arraybuffer'
import * as ImagePicker from 'expo-image-picker'
import React, { Ref, forwardRef, useImperativeHandle, useState } from 'react'

export interface UploadAvatarRefObject {
  pickImage: () => void
}

export const UploadAvatar = forwardRef(function UploadAvatar(
  { children }: { children: React.ReactNode },
  ref: Ref<UploadAvatarRefObject>
) {
  const { user, updateProfile } = useUser()
  const supabase = useSupabase()
  const [errMsg, setErrMsg] = useState('')

  useImperativeHandle(ref, () => ({ pickImage }))

  const pickImage = async () => {
    // No permissions request is necessary for launching the image library
    const result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [1, 1],
      quality: 1,
      base64: true,
    })

    await uploadImage(result)
  }

  const uploadImage = async (pickerResult: ImagePicker.ImagePickerResult) => {
    if (pickerResult.canceled) {
      return
    }
    if (!user) return
    const image = pickerResult.assets[0]
    if (!image) {
      setErrMsg('No image provided.')
      return
    }

    const base64Image = image.base64

    if (!base64Image) {
      setErrMsg('No image provided.')
      return
    }

    const base64Str = base64Image.includes('base64,')
      ? base64Image.substring(base64Image.indexOf('base64,') + 'base64,'.length)
      : base64Image
    const res = decode(base64Str)

    if (!(res.byteLength > 0)) {
      setErrMsg('ArrayBuffer is null')
      // console.error('ArrayBuffer is null')
      return null
    }
    const result = await supabase.storage
      .from('avatars')
      .upload(`${user.id}/${Number(new Date())}.jpeg`, res, {
        contentType: 'image/jpeg',
        upsert: true,
      })
    if (result.error) {
      setErrMsg(result.error.message)
      return
      // console.log(result.error)
      // throw new Error(result.error.message)
    }
    const publicUrlRes = await supabase.storage
      .from('avatars')
      .getPublicUrl(result.data.path.replace('avatars/', ''))

    const { error: update_error } = await supabase
      .from('profiles')
      .update({ avatar_url: publicUrlRes.data.publicUrl })
      .eq('id', user.id)
    if (update_error) {
      return
    }
    await updateProfile()
  }

  return (
    <YStack gap={'$4'}>
      <YStack
        position="relative"
        alignSelf="flex-start"
        flexShrink={1}
        onPress={() => pickImage()}
        cursor="pointer"
      >
        {children}
        <YStack
          position="absolute"
          left={0}
          right={0}
          top={0}
          bottom={0}
          jc="center"
          ai="center"
          zIndex={100}
        >
          <YStack
            backgroundColor="black"
            opacity={0.34}
            borderRadius="$3"
            position="absolute"
            left={0}
            right={0}
            top={0}
            bottom={0}
            $md={{ borderRadius: 12 }}
          />
          <YStack position="absolute" left={0} right={0} top={0} bottom={0} jc="center" ai="center">
            <IconRefresh color="primary" />
          </YStack>
        </YStack>
      </YStack>
      {errMsg ? <SizableText theme="red">{errMsg}</SizableText> : <></>}
    </YStack>
  )
})
