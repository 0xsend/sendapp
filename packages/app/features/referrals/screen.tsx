import { Card, CardHeader, Container, H3, H5, Paragraph, XStack, YStack } from '@my/ui'
import { IconSendToken } from 'app/components/icons'
import { shortAddress } from 'app/utils/NameDisplayUtils'
import { useUserReferralsCount } from 'app/utils/UseUserReferralsCount'
import { useDistributionBonusPoolShares } from 'app/utils/useDistributionBonusPoolShares'

const pageDescription =
  "Share your unique link, and when someone joins or makes a purchase using it, you earn rewards in SEND token. Our system tracks referrals, ensuring you receive rewards. It's a win-win as you support us and enjoy the benefits. Thanks for being part of our community!"

export const ReferralsScreen = () => {
  return (
    <Container>
      <YStack f={1} width={'100%'} py="$4" gap="$4">
        <XStack alignItems="center" width={'100%'} jc="space-between" gap="$6">
          <H3 fontWeight={'normal'}>Referral Rewards</H3>
          <H3 fontWeight={'bold'} color={'$gold10'}>
            Check Rewards
          </H3>
        </XStack>
        <Paragraph py={'$2'} size={'$4'}>
          {pageDescription}
        </Paragraph>
        <StatsCards />
        <RewardsHistory />
      </YStack>
    </Container>
  )
}

const StatsCards = () => {
  const { referralsCount, error: referralsCountError } = useUserReferralsCount()
  const { data } = useDistributionBonusPoolShares()
  const totalRewards = data?.reduce((acc, { bonus_pool_amount }) => bonus_pool_amount + acc, 0)
  return (
    <XStack flexWrap="wrap" ai="flex-start" jc="space-between" gap="$8" mb="$4">
      <Card f={1} w={'20%'} bc={'$background'}>
        <CardHeader>
          <Paragraph>Total Referrals</Paragraph>
        </CardHeader>
        <H5 pl="$4" pb="$3" fontWeight="600" size={'$7'}>
          {referralsCountError ? '?' : referralsCount || 0}
        </H5>
      </Card>
      <Card f={1} w={'20%'} bc={'$background'}>
        <CardHeader>
          <Paragraph>Transactions</Paragraph>
        </CardHeader>
        <H5 pl="$4" pb="$3" fontWeight="600" size={'$7'}>
          ?
        </H5>
      </Card>
      <Card f={1} w={'20%'} bc={'$background'}>
        <CardHeader>
          <Paragraph>Total Rewards</Paragraph>
        </CardHeader>
        <XStack ai="center" h="100%" pl="$4" pb="$3" gap="$2">
          <IconSendToken />
          <H5 pl="$4" pb="$3" fontWeight="600" size={'$7'}>
            {totalRewards}
          </H5>
        </XStack>
      </Card>
      <Card f={1} w={'20%'} bc={'$background'}>
        <CardHeader>
          <Paragraph>Claimable Rewards</Paragraph>
        </CardHeader>
        <H5 pl="$4" pb="$3" fontWeight="600" size={'$7'}>
          ?
        </H5>
      </Card>
    </XStack>
  )
}

const RewardsHistory = () => {
  return (
    <YStack f={1} width={'100%'} space="$4">
      <XStack alignItems="center" width={'100%'} jc="space-between" gap="$6">
        <H3 fontWeight={'normal'}>Referral History</H3>
        <H3 fontWeight={'bold'}>See All</H3>
      </XStack>
      <YStack f={1} width={'100%'} p="$6" bg="$background" space="$4" br="$8">
        <ReferralsHistoryHeader />
      </YStack>
    </YStack>
  )
}

const RewardsHistoryHeader = () => (
  <XStack gap="$3" ai="center" jc="space-between">
    <Paragraph w="12%" mb="0" size="$5" lineHeight="$4">
      #
    </Paragraph>
    <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
      Date
    </Paragraph>
    <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
      Wallet
    </Paragraph>
    <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
      Referral rewards
    </Paragraph>
    <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
      Status
    </Paragraph>
    <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="right">
      Details
    </Paragraph>
  </XStack>
)

const RewardsHistoryData = () => {
  const { data } = useDistributionBonusPoolShares()

  return data?.map(({ bonus_pool_amount, created_at, address, distribution_id }) => (
    <XStack key={created_at.toString()} gap="$3" ai="center" jc="space-between">
      <Paragraph w="12%" mb="0" size="$5" lineHeight="$4">
        {distribution_id}
      </Paragraph>
      <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
        {created_at.toLocaleDateString('en-US', {
          day: 'numeric',
          year: 'numeric',
          month: 'short',
        })}
      </Paragraph>
      <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
        {shortAddress(address)}
      </Paragraph>
      <XStack f={1} ai="center" w="12%" jc="center" mb="0" gap="$2">
        <IconSendToken />
        <Paragraph size="$5" lineHeight="$4" ta="center">
          {bonus_pool_amount}
        </Paragraph>
      </XStack>

      <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
        {'?'}
      </Paragraph>
      <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="right">
        Transaction
      </Paragraph>
    </XStack>
  ))
}
