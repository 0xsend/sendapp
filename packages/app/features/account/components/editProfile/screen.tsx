import {
  Button,
  Checkbox,
  FadeCard,
  Paragraph,
  ProfileAvatar,
  Separator,
  Spinner,
  SubmitButton,
  XStack,
  YStack,
} from '@my/ui'
import { SchemaForm } from 'app/utils/SchemaForm'
import { ProfileSchema, useProfileMutation } from 'app/utils/useProfileMutation'
import { useUser } from 'app/utils/useUser'
import { UploadAvatar, type UploadAvatarRefObject } from '../uploadProfileImage/screen'
import { useRef, useState } from 'react'
import type { Tables } from '@my/supabase/database.types'
import { Check } from '@tamagui/lucide-icons'
import { SettingsHeader } from 'app/features/account/settings/components/SettingsHeader'
import { ReadOnlyFieldWithLabel } from 'app/features/account/settings/components/ReadOnlyFieldWithLabel'
import { FieldWithLabel } from 'app/features/account/settings/components/FieldWithLabel'
import { useForm } from 'react-hook-form'
import type { z } from 'zod'

enum FormState {
  Overview = 'Overview',
  ProfileForm = 'ProfileForm',
}

export const EditProfile = () => {
  const [formState, setFormState] = useState<FormState>(FormState.Overview)
  const { profile } = useUser()

  return (
    <YStack w={'100%'}>
      <YStack gap={'$3.5'}>
        <SettingsHeader>Profile</SettingsHeader>
        {(() => {
          switch (true) {
            case !profile:
              return <Spinner size="large" />
            case formState === FormState.Overview:
              return (
                <Overview profile={profile} onPress={() => setFormState(FormState.ProfileForm)} />
              )
            case formState === FormState.ProfileForm:
              return (
                <EditProfileForm
                  profile={profile}
                  onSave={() => setFormState(FormState.Overview)}
                />
              )
          }
        })()}
      </YStack>
    </YStack>
  )
}

const Overview = ({ profile, onPress }: { profile: Tables<'profiles'>; onPress: () => void }) => {
  const { name, about, is_public, avatar_url } = profile
  const avatarRef = useRef<UploadAvatarRefObject>(null)

  return (
    <YStack gap={'$5'}>
      <FadeCard>
        <XStack gap={'$5'} width={'100%'}>
          <UploadAvatar ref={avatarRef}>
            <ProfileAvatar
              avatarUrl={avatar_url ? avatar_url : undefined}
              $gtMd={{ size: 88 }}
              size={88}
            />
          </UploadAvatar>
          <YStack gap={'$2'} ai={'flex-start'}>
            <YStack>
              <Paragraph
                size={'$5'}
                fontWeight={500}
                color={'$lightGrayTextField'}
                $theme-light={{ color: '$darkGrayTextField' }}
              >
                Profile Picture
              </Paragraph>
            </YStack>
            <Button unstyled onPress={() => avatarRef.current?.pickImage()}>
              <Button.Text
                textDecorationLine="underline"
                color="$primary"
                $theme-light={{ color: '$color12' }}
                size={'$5'}
              >
                Update
              </Button.Text>
            </Button>
          </YStack>
        </XStack>
        <FieldWithLabel label={'Name'}>
          <Paragraph size={'$8'} fontWeight={'500'}>
            {name || '-'}
          </Paragraph>
        </FieldWithLabel>
        <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
        <ReadOnlyFieldWithLabel label={'About'} text={about || '-'} />
        <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
        <XStack gap={'$3'} ai={'center'}>
          <Checkbox
            disabled={true}
            checked={!!is_public}
            borderWidth={0}
            backgroundColor={is_public ? '$primary' : '$background'}
            circular={true}
          >
            <Checkbox.Indicator>
              <Check color={'$black'} />
            </Checkbox.Indicator>
          </Checkbox>
          <Paragraph size={'$5'}>{`${is_public ? 'Public' : 'Private'}`} Profile</Paragraph>
        </XStack>
      </FadeCard>
      <SubmitButton theme="green" borderRadius={'$4'} p={'$4'} onPress={onPress}>
        <Button.Text ff={'$mono'} fontWeight={'500'} tt="uppercase" size={'$5'} color={'$black'}>
          edit profile
        </Button.Text>
      </SubmitButton>
    </YStack>
  )
}

function EditProfileForm({ profile, onSave }: { profile: Tables<'profiles'>; onSave: () => void }) {
  const { id, name, about, is_public } = profile
  const { mutateAsync, error } = useProfileMutation(id)
  const form = useForm<z.infer<typeof ProfileSchema>>()

  const handleSubmit = async () => {
    const values = form.getValues()
    await mutateAsync(values)
    onSave()
  }

  return (
    <SchemaForm
      form={form}
      schema={ProfileSchema}
      props={{
        name: {
          'aria-label': 'Name',
          bc: '$color0',
        },
        about: {
          'aria-label': 'Bio',
          placeholder: 'Tell us about yourself',
          backgroundColor: '$color0',
          rows: 2,
          focusStyle: {
            fontStyle: 'normal',
          },
        },
        isPublic: {
          defaultChecked: is_public !== null ? is_public : true,
        },
      }}
      defaultValues={{
        name: name ? name : '',
        about: about ? about : '',
        isPublic: is_public !== null ? is_public : true,
      }}
      formProps={{
        $gtSm: {
          maxWidth: '100%',
        },
      }}
      onSubmit={handleSubmit}
      renderAfter={({ submit }) => (
        <YStack>
          <SubmitButton
            theme="green"
            borderRadius={'$4'}
            p={'$4'}
            mt={'$1'}
            onPress={() => submit()}
          >
            <Button.Text
              ff={'$mono'}
              fontWeight={'500'}
              tt="uppercase"
              size={'$5'}
              color={'$black'}
            >
              SAVE CHANGES
            </Button.Text>
          </SubmitButton>
          {error && (
            <Paragraph marginTop={'$3'} theme="red" color="$color9">
              {error.message}
            </Paragraph>
          )}
        </YStack>
      )}
    >
      {({ name, about, isPublic }) => (
        <FadeCard>
          <FieldWithLabel label={'Name'} gap={'$2'}>
            {name}
          </FieldWithLabel>
          <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
          <FieldWithLabel label={'About'} gap={'$2'}>
            {about}
          </FieldWithLabel>
          <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
          <XStack gap={'$3'} ai={'center'}>
            {isPublic}
            <Paragraph size={'$5'}>Make Profile Public</Paragraph>
          </XStack>
        </FadeCard>
      )}
    </SchemaForm>
  )
}
