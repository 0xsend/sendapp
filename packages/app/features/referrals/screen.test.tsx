import { expect, test } from '@jest/globals'
import { TamaguiProvider, config } from '@my/ui'
import { render } from '@testing-library/react-native'
import { ReferralsScreen } from './screen'

jest.mock('app/utils/useUser', () => ({
  useUser: jest.fn().mockReturnValue({
    user: {
      id: '123',
      profile: { referral_code: '123' },
    },
  }),
}))

jest.mock('app/routers/params', () => ({
  useNav: jest.fn().mockReturnValue([undefined, jest.fn()]),
}))
jest.mock('app/utils/UseUserReferralsCount', () => ({
  useUserReferralsCount: jest.fn().mockReturnValue({ referralsCount: 123, error: null }),
}))

jest.mock('app/utils/useDistributionBonusPoolShares', () => ({
  useDistributionBonusPoolShares: jest
    .fn()
    .mockReturnValue([{ bonus_pool_amount: [123] }, { bonus_pool_amount: [456] }]),
}))

jest.mock('app/utils/NameDisplayUtils', () => ({
  shorten: jest.fn(),
}))

test('ReferralScreen', () => {
  const tree = render(
    <TamaguiProvider defaultTheme={'dark'} config={config}>
      <ReferralsScreen />
    </TamaguiProvider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
