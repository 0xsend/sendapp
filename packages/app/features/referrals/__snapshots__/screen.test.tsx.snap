// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferralScreen 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flex": 1,
      "flexDirection": "row",
      "paddingLeft": 32,
      "paddingRight": 32,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "flexDirection": "column",
        "paddingBottom": 18,
        "paddingTop": 18,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 32,
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <Text
        accessibilityRole="header"
        selectable={true}
        style={
          {
            "color": "hsla(112, 22%, 59%, 1)",
            "fontFamily": "System",
            "fontSize": 23,
            "fontWeight": "bold",
            "lineHeight": 27,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "textTransform": "none",
          }
        }
        suppressHighlighting={true}
      >
        Referral Rewards
      </Text>
      <Text
        accessibilityRole="header"
        selectable={true}
        style={
          {
            "color": "hsla(112, 22%, 59%, 1)",
            "fontFamily": "System",
            "fontSize": 23,
            "fontWeight": "bold",
            "lineHeight": 27,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "textTransform": "none",
          }
        }
        suppressHighlighting={true}
      >
        Check Rewards
      </Text>
    </View>
    <View
      pointerEvents="none"
      style={
        {
          "height": 18,
          "minHeight": 18,
          "minWidth": 18,
          "width": 18,
        }
      }
    />
    <Text
      selectable={true}
      style={
        {
          "color": "hsla(112, 22%, 59%, 1)",
          "fontFamily": "System",
          "fontSize": 16.8,
          "fontWeight": "400",
          "lineHeight": 19,
          "paddingBottom": 7,
          "paddingTop": 7,
        }
      }
      suppressHighlighting={true}
    >
      Share your unique link, and when someone joins or makes a purchase using it, you earn rewards in SEND token. Our system tracks referrals, ensuring you receive rewards. It's a win-win as you support us and enjoy the benefits. Thanks for being part of our community!
    </Text>
    <View
      pointerEvents="none"
      style={
        {
          "height": 18,
          "minHeight": 18,
          "minWidth": 18,
          "width": 18,
        }
      }
    />
    <View
      style={
        {
          "alignItems": "flex-start",
          "flexDirection": "row",
          "flexWrap": "wrap",
          "gap": 46,
          "justifyContent": "space-between",
          "marginBottom": 18,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "hsla(191, 32%, 15%, 1)",
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
            "flex": 1,
            "flexDirection": "column",
            "position": "relative",
            "width": "20%",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "transparent",
              "flexDirection": "column",
              "marginBottom": "auto",
              "paddingBottom": 18,
              "paddingLeft": 18,
              "paddingRight": 18,
              "paddingTop": 18,
              "zIndex": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "fontFamily": "System",
                "fontSize": 16.8,
                "fontWeight": "400",
                "lineHeight": 19,
              }
            }
            suppressHighlighting={true}
          >
            Total Referrals
          </Text>
        </View>
        <Text
          accessibilityRole="header"
          selectable={true}
          style={
            {
              "color": "hsla(112, 22%, 59%, 1)",
              "fontFamily": "System",
              "fontSize": 20,
              "fontWeight": "700",
              "lineHeight": 24,
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "paddingBottom": 13,
              "paddingLeft": 18,
              "textTransform": "none",
            }
          }
          suppressHighlighting={true}
        >
          1000
        </Text>
      </View>
      <View
        style={
          {
            "backgroundColor": "hsla(191, 32%, 15%, 1)",
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
            "flex": 1,
            "flexDirection": "column",
            "position": "relative",
            "width": "20%",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "transparent",
              "flexDirection": "column",
              "marginBottom": "auto",
              "paddingBottom": 18,
              "paddingLeft": 18,
              "paddingRight": 18,
              "paddingTop": 18,
              "zIndex": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "fontFamily": "System",
                "fontSize": 16.8,
                "fontWeight": "400",
                "lineHeight": 19,
              }
            }
            suppressHighlighting={true}
          >
            Transactions
          </Text>
        </View>
        <Text
          accessibilityRole="header"
          selectable={true}
          style={
            {
              "color": "hsla(112, 22%, 59%, 1)",
              "fontFamily": "System",
              "fontSize": 20,
              "fontWeight": "700",
              "lineHeight": 24,
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "paddingBottom": 13,
              "paddingLeft": 18,
              "textTransform": "none",
            }
          }
          suppressHighlighting={true}
        >
          1000
        </Text>
      </View>
      <View
        style={
          {
            "backgroundColor": "hsla(191, 32%, 15%, 1)",
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
            "flex": 1,
            "flexDirection": "column",
            "position": "relative",
            "width": "20%",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "transparent",
              "flexDirection": "column",
              "marginBottom": "auto",
              "paddingBottom": 18,
              "paddingLeft": 18,
              "paddingRight": 18,
              "paddingTop": 18,
              "zIndex": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "fontFamily": "System",
                "fontSize": 16.8,
                "fontWeight": "400",
                "lineHeight": 19,
              }
            }
            suppressHighlighting={true}
          >
            Total Rewards
          </Text>
        </View>
        <Text
          accessibilityRole="header"
          selectable={true}
          style={
            {
              "color": "hsla(112, 22%, 59%, 1)",
              "fontFamily": "System",
              "fontSize": 20,
              "fontWeight": "700",
              "lineHeight": 24,
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "paddingBottom": 13,
              "paddingLeft": 18,
              "textTransform": "none",
            }
          }
          suppressHighlighting={true}
        >
          1000
        </Text>
      </View>
      <View
        style={
          {
            "backgroundColor": "hsla(191, 32%, 15%, 1)",
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
            "flex": 1,
            "flexDirection": "column",
            "position": "relative",
            "width": "20%",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "transparent",
              "flexDirection": "column",
              "marginBottom": "auto",
              "paddingBottom": 18,
              "paddingLeft": 18,
              "paddingRight": 18,
              "paddingTop": 18,
              "zIndex": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "fontFamily": "System",
                "fontSize": 16.8,
                "fontWeight": "400",
                "lineHeight": 19,
              }
            }
            suppressHighlighting={true}
          >
            Claimable Rewards
          </Text>
        </View>
        <Text
          accessibilityRole="header"
          selectable={true}
          style={
            {
              "color": "hsla(112, 22%, 59%, 1)",
              "fontFamily": "System",
              "fontSize": 20,
              "fontWeight": "700",
              "lineHeight": 24,
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "paddingBottom": 13,
              "paddingLeft": 18,
              "textTransform": "none",
            }
          }
          suppressHighlighting={true}
        >
          1000
        </Text>
      </View>
    </View>
    <View
      pointerEvents="none"
      style={
        {
          "height": 18,
          "minHeight": 18,
          "minWidth": 18,
          "width": 18,
        }
      }
    />
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 32,
            "justifyContent": "space-between",
            "width": "100%",
          }
        }
      >
        <Text
          accessibilityRole="header"
          selectable={true}
          style={
            {
              "color": "hsla(112, 22%, 59%, 1)",
              "fontFamily": "System",
              "fontSize": 23,
              "fontWeight": "normal",
              "lineHeight": 27,
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "textTransform": "none",
            }
          }
          suppressHighlighting={true}
        >
          Referral History
        </Text>
        <Text
          accessibilityRole="header"
          selectable={true}
          style={
            {
              "color": "hsla(112, 22%, 59%, 1)",
              "fontFamily": "System",
              "fontSize": 23,
              "fontWeight": "bold",
              "lineHeight": 27,
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "textTransform": "none",
            }
          }
          suppressHighlighting={true}
        >
          See All
        </Text>
      </View>
      <View
        pointerEvents="none"
        style={
          {
            "height": 18,
            "minHeight": 18,
            "minWidth": 18,
            "width": 18,
          }
        }
      />
      <View
        style={
          {
            "backgroundColor": "hsla(191, 32%, 10%, 1)",
            "borderBottomLeftRadius": 22,
            "borderBottomRightRadius": 22,
            "borderTopLeftRadius": 22,
            "borderTopRightRadius": 22,
            "flex": 1,
            "flexDirection": "column",
            "paddingBottom": 32,
            "paddingLeft": 32,
            "paddingRight": 32,
            "paddingTop": 32,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "gap": 13,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            selectable={true}
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "fontFamily": "System",
                "fontSize": 19.2,
                "fontWeight": "400",
                "lineHeight": 19,
                "marginBottom": "0",
                "width": "12%",
              }
            }
            suppressHighlighting={true}
          >
            #
          </Text>
          <Text
            selectable={true}
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "flex": 1,
                "fontFamily": "System",
                "fontSize": 19.2,
                "fontWeight": "400",
                "lineHeight": 19,
                "marginBottom": "0",
                "textAlign": "center",
                "width": "12%",
              }
            }
            suppressHighlighting={true}
          >
            Date
          </Text>
          <Text
            selectable={true}
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "flex": 1,
                "fontFamily": "System",
                "fontSize": 19.2,
                "fontWeight": "400",
                "lineHeight": 19,
                "marginBottom": "0",
                "textAlign": "center",
                "width": "12%",
              }
            }
            suppressHighlighting={true}
          >
            Wallet
          </Text>
          <Text
            selectable={true}
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "flex": 1,
                "fontFamily": "System",
                "fontSize": 19.2,
                "fontWeight": "400",
                "lineHeight": 19,
                "marginBottom": "0",
                "textAlign": "center",
                "width": "12%",
              }
            }
            suppressHighlighting={true}
          >
            Referral rewards
          </Text>
          <Text
            selectable={true}
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "flex": 1,
                "fontFamily": "System",
                "fontSize": 19.2,
                "fontWeight": "400",
                "lineHeight": 19,
                "marginBottom": "0",
                "textAlign": "center",
                "width": "12%",
              }
            }
            suppressHighlighting={true}
          >
            Status
          </Text>
          <Text
            selectable={true}
            style={
              {
                "color": "hsla(112, 22%, 59%, 1)",
                "flex": 1,
                "fontFamily": "System",
                "fontSize": 19.2,
                "fontWeight": "400",
                "lineHeight": 19,
                "marginBottom": "0",
                "textAlign": "right",
                "width": "12%",
              }
            }
            suppressHighlighting={true}
          >
            Details
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
